File: src/subroutine/domain/journal.py
"""What happened over a period, as against what changed — item `#1430`, decision `#1429`.

**One store, two reads.** :mod:`subroutine.domain.events` answers *what changed*: raw, cheap,
complete, resumable from a cursor, and exactly what a client polling for work wants. This
answers *what happened*, which is a different question asked by a person or by an agent told to
say what a stretch of time contained — and the difference between the two is entirely a
**join**.

**Measured before it was designed**, on one day of real work: 450 events, of which 130 were
``comment.created`` carrying no body at all, 51 field-changes whose values were bare UUIDs, and
an actor column that was a UUID on every single row. So the feed had the skeleton — ``seq``, an
order, and a title on every row — and none of the substance.

**Nothing is written differently.** The obvious fix is to put the comment's body on the event,
and it is refused: ``event.changes`` already stores ``from`` and ``to`` in full rather than as a
diff, in a table that is never pruned, which is a filed bug at `#578`. Copying bodies in would
compound it on 29% of the feed. The events already carry the ids; a second reader joins.
"""

import typing
import uuid

import sqlalchemy
import sqlalchemy.orm

import subroutine.db.models.activity
import subroutine.db.models.work
import subroutine.domain.authentication
import subroutine.domain.events
import subroutine.domain.scoping
import subroutine.domain.text

#: The lookups a value can be named by. Strings rather than an enum because they are keys into
#: :class:`subroutine.views.Vocabulary`, which is where the batch loading already lives.
USER = "user"
STATUS = "status"
TYPE = "type"
PROJECT = "project"
TASK = "task"

#: Which lookup names the value of an id-valued change, by the column it moved.
#:
#: **Declared rather than inferred, because a name cannot answer it.** A column ending ``_id``
#: says nothing about which table it points at, and guessing from the stem would put
#: ``recurrence_template_id`` at a ``recurrence_template`` table that does not exist.
#:
#: **What is missing degrades to silence rather than to a UUID**, which is the design decision
#: worth stating: a column absent from here renders its phrase and no value — *changed how it
#: repeats* — where a UUID would be noise a reader has to learn to skip. So a column added
#: tomorrow is unhelpful here rather than wrong, and
#: ``test_no_journal_entry_ever_renders_an_identifier`` is what says nobody has quietly started
#: relying on the other behaviour.
NAMED_BY: dict[str, str] = {
	"assignee_id": USER,
	"assigned_by_id": USER,
	"claimed_by_id": USER,
	"owner_id": USER,
	"created_by": USER,
	"updated_by": USER,
	"status_id": STATUS,
	"type_id": TYPE,
	"project_id": PROJECT,
	"parent_task_id": TASK,
	"recurrence_template_id": TASK,
}


#: The fields whose value is a whole text, on any kind of thing — `#2728`. **A journal entry
#: never carries one** (Simon, 2026-09-16): a change to any of these says that it changed and
#: nothing either side, and the audit log keeps both sides whole for anybody who needs them.
#:
#: **By field name rather than per entity**, and wider than
#: :data:`subroutine.domain.events.PROSE_FIELD` on purpose. That one names the field whose
#: replacement counts as a *revision*, one per kind; this is every field that holds prose at all,
#: which is also a comment's body when it is edited, a project's description and a workspace's.
#: Measured on the served instance's latest 5,000 events: the text inside their changes was
#: 5.2 MB of documents' bodies and 0.7 MB of tasks' descriptions, against 0.2 MB for every
#: other field together.
#: ``test_every_prose_field_is_a_whole_text`` holds the one inside the other.
#:
#: **A title is one too, for the journal's reason rather than for its size** (Simon,
#: 2026-09-17, `#2853`). The item's row already shows the title it has now, so the two either
#: side of *to* said nothing a reader needed - and they made the longest lines: on this
#: instance's latest 400 entries, all six title changes were over 140 characters, up to 367,
#: where every kind of line but a comment was at most 53.
WHOLE_TEXTS: frozenset[str] = frozenset({"description", "body", "title"})

#: How much of a comment an entry carries — `#2728`. It was 280, measured to keep most first
#: paragraphs whole, and **Simon halved it on 2026-09-17** (`#2852`) when line breaks became
#: spaces: on the latest 48 comments in the journal 23 reached 280, about three lines of the
#: page each, and 140 keeps the first sentence - the heading an agent writes first - in one
#: and a half.
OPENING = 140


class Said(typing.NamedTuple):
	"""The opening of what a comment said, and whether there was more of it."""

	opening: str
	cut: bool


def _not_a_rule_bearing_row () -> typing.Any:
	"""Refuse an entry about the hidden row a repeat keeps its rule on - `#2849`.

	**One act by a person is one entry** (Simon, 2026-09-20). Giving an item a repeat makes a
	second row that holds the rule, in no listing and reachable only by a number nobody was
	shown, and the journal printed its creation above the change that made it: *created #2
	Water the plants*, then *updated #1 Water the plants*. A reader met a twin of the item
	they were looking at, and nothing said what it was. Decision `#1249`'s framing is the
	argument: a repeating item is **one** thing to the person who filed it.

	**Nothing is lost by leaving it out.** The act itself is recorded against the item a
	reader can reach (`#2825`), the series' own history is on its own page for anybody holding
	its number, and the events are untouched - so ``/v1/changes``, the feed a client polls,
	still carries them. That is the split `#1429` made: one store, two readings.

	**The same flag every listing already uses**, ``task.is_template``, rather than a second
	description of what a series row looks like.

	**Asked of each event's own task, by its key** (the cold review of 2026-09-21, `#3159`). This
	was ``entity_id IN (every template in the installation)``, unnarrowed, over a column with no
	index - so each page, and the browser polls pages, read the whole task table to take a
	handful of rows out of a page of fifty. A correlated ``EXISTS`` looks up one task by its
	primary key for each event about a task, and grows with the page rather than the install.
	"""

	event = subroutine.db.models.activity.Event
	task = subroutine.db.models.work.Task

	return ~(
		(event.entity_type == TASK)
		& sqlalchemy.exists().where(task.id == event.entity_id, task.is_template.is_(True))
	)


def page (
	session: sqlalchemy.orm.Session,
	principal: subroutine.domain.authentication.Principal,
	*,
	workspace_ids: typing.Sequence[uuid.UUID],
	size: int,
	mine: bool = False,
	by: uuid.UUID | None = None,
	oldest: bool = False,
	narrowing: typing.Sequence[typing.Any] = (),
) -> tuple[list[subroutine.db.models.activity.Event], bool]:
	"""Return one page of the journal and whether more lie beyond it — `#2772`.

	**The latest entries, newest first**, or with ``oldest`` the first of a period, oldest first.
	Simon's decision of 2026-09-16, and his reason is the next page: read from the newest, the
	page after this one is the entries before it, and read from the oldest it is the entries
	after. The last hundred oldest first would leave the next page meaning nothing.

	**Here rather than in each transport**, because the rows are :func:`events.page`'s and that
	function always answers forwards, which is right for the feed a cursor resumes. Two routes
	reversing it for themselves is two answers to one question waiting to disagree.

	**And a repeat's rule-bearing row is left out here rather than in the feed** (`#2849`),
	because it is this reading of the events that has a person in front of it.
	:func:`_not_a_rule_bearing_row` carries the argument.
	"""

	rows, more = subroutine.domain.events.page(
		session,
		principal,
		workspace_ids=workspace_ids,
		size=size,
		mine=mine,
		by=by,
		newest=not oldest,
		# A condition beside the caller's filters rather than one of them: both are clauses
		# this statement is narrowed by, and only this one is the journal's own rule.
		narrowing=(*narrowing, _not_a_rule_bearing_row()),
	)

	if not oldest:
		rows.reverse()

	return rows, more


def identifier (value: typing.Any) -> uuid.UUID | None:
	"""Return ``value`` as an id if that is what it is, and ``None`` otherwise.

	``event.changes`` is JSON, so what comes back is a string on both backends rather than the
	``uuid.UUID`` the column held. Parsing is the check: a value that is not one raises and is
	simply not an id, which is the same answer as a column nobody declared.
	"""

	if isinstance(value, uuid.UUID):
		return value

	if not isinstance(value, str):
		return None

	try:
		return uuid.UUID(value)

	except ValueError:
		return None


def wanted (
	rows: typing.Sequence[subroutine.db.models.activity.Event],
	described: typing.Mapping[uuid.UUID, subroutine.domain.events.Described],
) -> dict[str, set[uuid.UUID]]:
	"""Return which ids each lookup must fetch to render this page, keyed by lookup.

	**Both sides of every change**, because a journal says what something moved *from* as well
	as what it moved to — and *In progress to Done* is the sentence, where *to Done* leaves a
	reader to remember what it was.

	**Actors are in here too**, under :data:`USER`. They are the one id on an event that is not
	inside ``changes``, and forgetting them is how a page resolves every status perfectly and
	still says a UUID did it.

	**And what each entry's item is and where it is filed** (`#2727`), out of ``described``,
	which loaded them with the item's ref and title. Under the same lookups as a change's, so
	the project an item is filed in passes :func:`readable_only` like every other project an
	entry names, rather than being the one name on the entry that skipped it.
	"""

	found: dict[str, set[uuid.UUID]] = {
		USER: set(), STATUS: set(), TYPE: set(), PROJECT: set(), TASK: set()
	}

	for about in described.values():
		if about.type_id is not None:
			found[TYPE].add(about.type_id)

		if about.project_id is not None:
			found[PROJECT].add(about.project_id)

	for row in rows:
		if row.actor_user_id is not None:
			found[USER].add(row.actor_user_id)

		if not isinstance(row.changes, dict):
			continue

		for field, moved in row.changes.items():
			lookup = NAMED_BY.get(field)

			if lookup is None or not isinstance(moved, dict):
				continue

			for side in ("from", "to"):
				# **Not `identifier`**, which is the name of the function two lines up — and
				# assigning to it here shadowed it for the rest of the loop, so the *second*
				# side of the first change raised. `#1409`'s defect: a name bound twice, and
				# Python taking the later binding.
				found_id = identifier(moved.get(side))

				if found_id is not None:
					found[lookup].add(found_id)

	return found


def readable_only (
	session: sqlalchemy.orm.Session,
	needed: dict[str, set[uuid.UUID]],
	*,
	principal: subroutine.domain.authentication.Principal,
	workspace_ids: typing.Sequence[uuid.UUID],
) -> dict[str, set[uuid.UUID]]:
	"""Keep only the ids inside a page's changes that its reader may see named — `#2726`.

	**An entry can be visible while something named inside it is not.** Whether a reader gets an
	entry at all is decided by the item it is about; what moved *inside* the change is another
	row. A task moved out of a private project into an open one is rightly shown to somebody
	outside the private project, and its entry said *where it is filed: 'secret' to 'open'* to
	them - measured on both backends before this existed.

	**A project and a task are the lookups that can name something private.** An account, a
	status and a type belong to the workspace, and everybody reading it may know them. What is
	dropped here renders as nothing, exactly as an id nobody can name already does.
	"""

	kept = dict(needed)

	for lookup in (PROJECT, TASK):
		if not needed[lookup]:
			continue

		statement = subroutine.domain.scoping.readable_among(
			principal, workspace_ids=workspace_ids, kind=lookup, identifiers=needed[lookup]
		)

		# A kind this credential cannot read at all names nothing of that kind.
		kept[lookup] = set() if statement is None else set(session.scalars(statement))

	return kept


def said (
	session: sqlalchemy.orm.Session,
	rows: typing.Sequence[subroutine.db.models.activity.Event],
) -> dict[uuid.UUID, Said]:
	"""Return how each comment on this page opens, keyed by the comment's own id.

	**The one thing the feed omits and the whole reason this module exists.** A
	``comment.created`` event names the comment as its entity and says nothing about its
	contents, so 29% of a day's feed is a row reporting only that somebody wrote something.

	**A deleted comment is absent rather than empty.** Deletion is soft, so the row and its body
	are both still there — and showing them would make the journal the one surface where a
	retracted paragraph is still readable. Absent is the same answer the mention index gives:
	a deleted comment stops mentioning anything, because a backlink to a sentence nobody can
	read is worse than none.

	One query for the page, whatever its size, which is `#39`'s rule and the reason
	:func:`subroutine.domain.events.descriptions` next door is shaped the same way.

	**Only the opening, and cut here rather than by whoever renders it** (`#2728`), so no whole
	comment leaves this function: :data:`OPENING` characters on one line, ended at a word unless
	that would keep fewer than a third of them, and marked.
	"""

	model = subroutine.db.models.activity.Comment
	wanted_ids = {
		row.entity_id
		for row in rows
		if row.entity_type == "comment" and row.entity_id is not None
	}

	if not wanted_ids:
		return {}

	found = session.execute(
		sqlalchemy.select(model.id, model.body).where(
			model.id.in_(wanted_ids), model.deleted_at.is_(None)
		)
	).tuples()

	return {
		identifier: Said(*subroutine.domain.text.opening(body, OPENING))
		for identifier, body in found
		if body
	}

File: src/subroutine/web/assets/journal.js
/*
	What happened, in a workspace or to one item — `#2731` and `#1428`, design `#2724`.

	**A page drawn in place of the work, as the settings and people pages are**, and for their
	reason: it has no arrangement or selection of its own, so it is not a fourth view. It is
	reached at `/<workspace>/-/journal` and `/<workspace>/<ref>/-/journal` (`journalPageOf`).

	**Everything here is a rendering of what `/v1/journal` and an item's journal already say.**
	Who did each thing, through which door and never with which credential, what a change moved
	between, and how a comment opens. No entry carries a whole text (`#2728`), so nothing here
	renders prose as Markdown: an opening is a sentence cut at a word, and the item has the rest.

	**Drawn the way the list draws an item** (`#2826`, Simon's): each run of entries about one
	item is that item's row, with a line under it for each thing that happened, so the
	project's colour and the item's status are scanned without reading.

	Hook-free, so the render harness can call every component here (`#640`).
*/

import { html } from "./html.js";
import { addressOf, journalAddress } from "./address.js";
import { day } from "./dates.js";
import { clock } from "./marks.js";
import { Row } from "./rows.js";

/*
	**The door a change came in through, in words** — `#2727`. The instance records which door;
	it never records a credential's title here, because a title is whatever its owner chose to
	call it and says something about a colleague's own setup (Simon, 2026-09-16).

	A door this page does not know is left unsaid rather than printed raw: a new door is a word
	to add here, and a key on the page would be a reader decoding a program.
*/
export const DOORS = {
	browser: "in the browser",
	api: "through the API",
	mcp: "through agent tools",
	feed: "from a calendar feed",
	local: "on the instance's own machine",
};

export function mergedEntries (held, arriving) {
	/*
		The entries a page holds once more arrive — newest first, and each once.

		**Keyed by `seq`**, the number the instance gives every event. A page reads from the edge
		of what it holds, inclusively, so the entry at the edge comes back and is dropped here
		rather than drawn twice. Inclusive because two entries can share an instant, and a strict
		bound would lose the second at a page boundary.

		**Sorted rather than trusted.** Both journals answer newest first (`#2772`), and a newer read
		lands above what the page holds while an older one lands below it, so the order is the
		page's to keep rather than any one answer's.
	*/
	const found = new Map();

	for (const entry of [...(held || []), ...(arriving || [])]) found.set(entry.seq, entry);

	return [...found.values()].sort((one, other) => other.seq - one.seq);
}

export function journalBounds (holding, direction) {
	/*
		Where a read starts, given what the page holds — `#2731`, `#1428`.

		**Nothing on arrival.** **The newest entry held, for what is newer**, and **the oldest, for
		what is older** — each an instant, which `journalRequest` sends inclusively. **An item's
		cursor, for older**, because its route pages by cursor rather than by a period.

		Pure, so the decision is checkable without a mounted page (`#640`): a wrong edge here reads
		the same page again or skips one, and looks like nothing at all.
	*/
	const held = (holding && holding.entries) || [];

	return {
		from: direction === "newer" && held.length > 0 ? held[0].created_at : null,
		until: direction === "older" && held.length > 0 ? held[held.length - 1].created_at : null,
		cursor: direction === "older" && holding ? holding.cursor || null : null,
	};
}

export function journalAfter (holding, direction, {
	address, arriving = [], more = false, cursor = null, kind = null,
}) {
	/*
		What the page holds once a read lands — `#2731`, `#1428`.

		**On arrival, what arrived.** **Newer and older are merged into what was held**, each once
		and newest first. **Older says whether there is further back**, from the read that just
		went there; a newer read leaves that answer as it was.

		**A newer read that could not reach what the page held has a gap behind it**: more arrived
		than one read carries, and none of it meets the newest entry already drawn. The page then
		starts again from what arrived, with older to ask for, rather than drawing two runs as if
		they were one.
	*/
	const held = (holding && holding.entries) || [];
	const reaches = held.length === 0 || arriving.some((entry) => entry.seq <= held[0].seq);
	const gap = direction === "newer" && more && !reaches;
	const kept = direction === "newer" && !gap;

	return {
		address,
		kind: kind || (holding && holding.kind) || null,
		entries: mergedEntries(direction === null || gap ? [] : held, arriving),
		older: kept ? Boolean(holding && holding.older) : more,
		cursor: kept ? (holding && holding.cursor) || null : cursor,
	};
}

export function byDay (entries) {
	/*
		Entries grouped under the day each happened, in the reader's own zone and in the order
		given.

		**The reader's zone, because these are instants the program recorded**, not days somebody
		wrote: `day` with no zone is *when was this, where I am*, which is the question.
	*/
	const days = [];

	for (const entry of entries || []) {
		const heading = day(entry.created_at);
		const last = days[days.length - 1];

		if (last && last.day === heading) {
			last.entries.push(entry);
		} else {
			days.push({ day: heading, entries: [entry] });
		}
	}

	return days;
}

/*
	What a comment's entry says happened, in `views._HAPPENED`'s words — the verb alone, because
	the row above names the item it was written on.
*/
const COMMENTED = {
	created: "commented",
	updated: "edited a comment",
	deleted: "deleted a comment",
};

/*
	The one action whose line is not its verb and *it* - `#2862`, Simon: *released it* reads as a
	release of code, where this is a lease being given up. `views._AN_ACTION` carries the same
	phrase for the terminal and the agent tools, and *claimed it* is left as it was.
*/
const HELD = {
	released: "released the claim",
};

export function changeInWords (change) {
	/*
		One change as its line says it — `#2826`, Simon's example: *changed status*, then what it
		moved between.

		**`views.change_in_words` in the browser's words** (decision `#2823`): an empty side as
		*never*, *nobody* or *nothing*, a chosen name in quotes, and a date in the reader's own
		locale. **A change with neither side named is its name alone**, which is what a text
		change always is here (`#2728`) and what an id nothing can name becomes (`#1430`). A test
		drives this and the server's renderer over the same changes.
	*/
	if (change.before === null && change.after === null) return `changed ${change.said}`;

	return `changed ${change.said} from ${sideInWords(change, change.before)} to ${
		sideInWords(change, change.after)}`;
}

function sideInWords (change, value) {
	/* One side of a change as it is read: its empty word, a day, quoted, or as it is. An
	   instance older than the words sends no `empty`, and every empty side was *nothing*. */
	if (value === null || value === undefined) return change.empty || "nothing";

	if (change.dated) return dayInWords(value);

	return change.quoted ? `"${value}"` : value;
}

function dayInWords (written) {
	/* A journal's date arrives as `2026-09-18` or `2026-09-18T17:00`, **already in the item's
	   own zone**, so the day is written from its text and never converted — `day` returns a
	   bare date untouched for exactly that reason — and the time is the text after it. */
	const text = String(written);

	if (!/^\d{4}-\d{2}-\d{2}/.test(text)) return text;

	const shown = day(text.slice(0, 10));

	return text.length > 10 ? `${shown}, ${text.slice(11, 16)}` : shown;
}

export function linkOf (entry) {
	/*
		What a link entry joined, from this item's side: the other item, the kind of link, and
		whether it was made or undone — `#2826`.

		**The other end, whichever end this is.** An entry about a link names the item it was made
		from, and read through the other item it names that one. **Null for anything it cannot
		read**, so the line says what happened rather than guessing at a number.
	*/
	if (entry.entity_type !== "link") return null;

	const made = entry.action !== "deleted";
	const side = (field) => {
		const change = (entry.changed || []).find((one) => one.field === field);

		return change ? (made ? change.after : change.before) : null;
	};
	const source = Number(side("source"));
	const target = Number(side("target"));
	const other = source === entry.item_ref ? target : source;
	const type = side("link_type");

	if (!Number.isInteger(other) || other <= 0 || !type) return null;

	return { made, other, type: String(type).replaceAll("_", " ") };
}

export function linesOf (entry) {
	/*
		What one entry adds under its item's row, a line a thing that happened — `#2826`.

		**The row names the item, so a line never does**: *commented*, *changed status from…*,
		*linked it to #2803*. **A line a change** for an update, because each is its own fact.
		Creating something is *created it*, whatever columns it was born with: drawing those would
		bury the page under what every new item looks like.
	*/
	const base = {
		at: entry.created_at,
		actor: entry.actor || "the instance",
		door: entry.actor_interface || null,
	};

	if (entry.entity_type === "comment") {
		return [{
			...base,
			text: COMMENTED[entry.action] || `${verbOf(entry)} a comment`,
			said: entry.said || null,
			cut: Boolean(entry.said_truncated),
		}];
	}

	const link = linkOf(entry);

	if (link) return [{ ...base, link }];

	if (entry.action === "updated" && (entry.changed || []).length > 0) {
		return entry.changed.map((change) => ({ ...base, text: changeInWords(change) }));
	}

	return [{ ...base, text: HELD[entry.action] || `${verbOf(entry)} it` }];
}

function verbOf (entry) {
	/* An action as a word: `created`, `claimed`, `moved`. */
	return String(entry.action || "").replaceAll("_", " ");
}

export function collapsed (lines) {
	/*
		A row's lines with adjacent repeats said once — Simon's rule, 2026-09-17.

		**Identical lines** — the same person, through the same door, saying the same thing —
		**become one, with the newest time and how many times.** **Links of one kind, made or
		undone together by one person, become one line naming every item**, in the order they
		were joined: an item linked to nine others drew nine lines that looked the same. **Only
		adjacent lines**, so a claim, a change and a claim again stay three, because the order is
		what the history is read by.

		Lines arrive newest first, so the line already kept is the newer and its time stands.
	*/
	const kept = [];

	for (const line of lines || []) {
		const last = kept[kept.length - 1];
		const alongside = Boolean(last) && last.actor === line.actor && last.door === line.door;

		if (alongside && last.link && line.link
			&& last.link.made === line.link.made && last.link.type === line.link.type) {
			if (last.link.others.includes(line.link.other)) {
				last.count += 1;
			} else {
				last.link.others.unshift(line.link.other);
			}

			continue;
		}

		if (alongside && !last.link && !line.link
			&& last.text === line.text && last.said === line.said) {
			last.count += 1;

			continue;
		}

		kept.push(line.link
			? { ...line, count: 1, link: { ...line.link, others: [line.link.other] } }
			: { ...line, count: 1 });
	}

	return kept;
}

export function journalRows (entries) {
	/*
		A day's entries as rows, **consecutive entries about one item sharing its row** — `#2826`,
		Simon's. A row is the item; its lines are what happened to it, newest first.

		**Consecutive, not every entry about the item**, so the page still reads in the order
		things happened: an item touched this morning and again this afternoon, with something else
		between, is two rows. **An entry about no item** — a workspace being made — is a row of its
		own, named by what the journal calls it.
	*/
	const rows = [];

	for (const entry of entries || []) {
		const last = rows[rows.length - 1];
		const about = entry.item_ref
			? `#${entry.item_ref}`
			: `${entry.entity_type}:${entry.item_title || ""}`;

		if (last && last.about === about) {
			last.lines.push(...linesOf(entry));
		} else {
			rows.push({
				about,
				seq: entry.seq,
				ref: entry.item_ref || null,
				title: entry.item_title || null,
				project: entry.item_project_path || null,
				lines: linesOf(entry),
			});
		}
	}

	return rows.map((row) => ({ ...row, lines: collapsed(row.lines) }));
}

function Line ({ line, workspace, address = null }) {
	/*
		One thing that happened, under its item's row: when, who, what — and the door it came
		through and how many times, quiet at the end, because they are context rather than news.
	*/
	const door = line.door ? DOORS[line.door] : null;
	/* A statement of its own rather than inline in the markup: the words end in *from*, and the
	   served-module scan reads *from* and a quote after it as an import. */
	const joined = line.link && line.link.made ? "linked it to" : "unlinked it from";
	const others = line.link
		? line.link.others.map((other, index) => html`${index === 0 ? "" : ", "}<a
			href=${addressOf({ ref: other }, workspace)}>#${other}</a>`)
		: null;

	return html`
		<li>
			<span class="clock">${clock(line.at)}</span>
			<span class="what">
				<strong>${line.actor}</strong>${" "}${line.link
					? html`${joined}${" "}${others}${
						" "}(${line.link.type})`
					: line.text}${line.said
					? html`${" "}<span class="said">"${line.said}${line.cut ? "…" : ""}"</span>${
						line.cut && address ? html`${" "}<a class="rest" href=${address}>more</a>` : null}`
					: null}${line.count > 1
					? html`<span class="door"> · ${line.count} times</span>`
					: null}${door ? html`<span class="door"> · ${door}</span>` : null}
			</span>
		</li>
	`;
}

function JournalRow ({ row, item = null, workspace, onGo = null }) {
	/*
		One item's row with what happened to it underneath — `#2826`.

		**The list's own row where the item could be read**, so the project's colour, the status
		and the marks are scanned the way the list is, and a line takes the row's colour bar by
		being inside it. **No Complete button and nothing opened in place**: this is a record, and
		its links are ordinary links to the item's page.

		**A plain row where it could not** — deleted since, or about something with no row of its
		own, like a workspace — drawn from what the journal named.
	*/
	const filed = row.project;
	const address = !row.ref
		? null
		: item
		? addressOf(item, workspace)
		: addressOf({ ref: row.ref, project_key: filed, project_path: filed }, workspace);
	const lines = html`
		<ul class="happened">
			${row.lines.map((line, index) => html`
				<${Line} key=${index} line=${line} workspace=${workspace} address=${address} />
			`)}
		</ul>
	`;

	if (item) {
		return html`
			<${Row} item=${item} workspace=${workspace} place=${{ workspace, project: null }}
				onGo=${onGo}>${lines}<//>
		`;
	}

	const identity = html`
		${row.ref ? html`<span class="stamp"><span class="ref">#${row.ref}</span></span>` : null}
		<span class="title">${row.title}</span>
	`;

	return html`
		<li>
			${address
				? html`<a class="row" href=${address}>${identity}</a>`
				: html`<span class="row">${identity}</span>`}
			${lines}
		</li>
	`;
}

function ReadInstead ({ workspaces }) {
	/*
		The journals a reader can open from a page that shows none, as links — `#2773`.

		**Their real addresses, never a pattern of one.** This said `/workspace/journal`, which
		reads as a path to type, and the address it was first seen refusing was that pattern with
		a workspace put in. Nothing where the reader has no workspace to offer.
	*/
	const held = workspaces || [];

	if (held.length === 0) return null;

	return html`Open the journal for${" "}${held.map((one, index) => html`${
		index === 0 ? "" : index === held.length - 1 ? " or " : ", "
	}<a href=${journalAddress({ workspace: one.slug, ref: null })}>${one.title}</a>`)}.`;
}

export function Journal ({
	page = null, journal = null, workspaces = [], address = null, onOlder = null, busy = false,
	items = {}, onGo = null,
}) {
	/*
		The page: what happened, newest first, a day at a time.

		**`journal` is the answer for the page at `address`**, and an answer for another page is
		drawn as *Reading…* rather than drawn here, so a journal left behind by the last page is
		never shown under this one's heading.

		**Older is asked for, never fetched ahead**: the latest hundred are what somebody opening
		this is looking for, and each page further back is a request they chose.

		**`items` are the rows, by number**, read beside the journal (`journalItemsRequests`);
		a number not in it draws a plain row from what the journal named.
	*/
	if (!page) {
		return html`
			<div class="journal">
				<h2 class="area">Journal</h2>
				<p class="empty">
					There is no journal at this address. Each workspace has one, and so does each item in it.${
					" "}<${ReadInstead} workspaces=${workspaces} />
				</p>
			</div>
		`;
	}

	const space = (workspaces || []).find((one) => one.slug === page.workspace);

	if ((workspaces || []).length > 0 && !space) {
		return html`
			<div class="journal">
				<h2 class="area">Journal</h2>
				<p class="empty">
					There is no workspace called ${page.workspace} that you can see.${
					" "}<${ReadInstead} workspaces=${workspaces} />
				</p>
			</div>
		`;
	}

	const current = journal && journal.address === address ? journal : null;
	const entries = current ? current.entries || [] : [];
	const first = entries.find((entry) => entry.item_ref === page.ref);
	const home = `/${encodeURIComponent(page.workspace)}`;

	return html`
		<div class="journal">
			<h2 class="area">Journal</h2>
			<p class="about">
				${page.ref === null
					? html`What has happened in the <a href=${home}>${space ? space.title : page.workspace}</a>${
						" "}workspace, newest first.`
					: html`What has happened to <a href=${`${home}/${page.ref}`}>#${page.ref}${first
						? ` ${first.item_title}` : ""}</a>, newest first.`}
			</p>
			${!current
				? html`<p class="empty">Reading…</p>`
				: current.failed
				? html`<p class="empty">This journal could not be read. ${current.failed}</p>`
				: entries.length === 0
				? html`<p class="empty">Nothing has happened here that you can see.</p>`
				: html`
					${byDay(entries).map((group) => html`
						<h3 class="day">${group.day}</h3>
						<ul class="rows">
							${journalRows(group.entries).map((row) => html`
								<${JournalRow} key=${row.seq} row=${row} item=${row.ref ? items[row.ref] || null : null}
									workspace=${page.workspace} onGo=${onGo} />
							`)}
						</ul>
					`)}
					${current.older && onOlder
						? html`<p class="older">
								<button type="button" class="action" disabled=${busy}
									onClick=${onOlder}>Older</button>
							</p>`
						: null}
				`}
		</div>
	`;
}
